Added tutorial page on log trace in application (#80)
MicroEJ/docs: Tutorials/tutorialTracesInApplication.rst

      .. code-block:: java

         private static final Tracer tracer = new Tracer("Application", 100);
      
      In this case, ``Application`` identifies a group of events that defines a maximum of ``100`` available event types.

#. Next, activate the trace system:

      .. code-block:: java

         public static void main(String[] args) {
            Tracer.startTrace();

            currentState = ApplicationState.UNINSTALLED;
            switchState(ApplicationState.INSTALLED);
         }

#. Use the methods Tracer.recordEvent(...) and Tracer.recordEventEnd(...) to record the entry/exit events in the method:

      .. code-block:: java

         public static void switchState(ApplicationState newState) {
            tracer.recordEvent(0);

            oldState = currentState;
            currentState = newState;

            tracer.recordEventEnd(0);
         }
   
   The ``Tracer`` object records the entry/exit of method ``switchState`` with event ID ``0``.
   
This will produce the following output:

      .. code-block::

         [TRACE: Application] Event 0x0()
         [TRACE: Application] Event End 0x0()

.. note::

>    The default platform implementation of the Trace API prints the events in the console. MicroEJ provides an other implementation that redirects the events to `SystemView <https://www.segger.com/products/development-tools/systemview/>`_, the real-time recording and visualization tool from `Segger <https://www.segger.com/>`_. It allows for a finer understanding of the runtime behavior by showing events sequencing and duration. A platform reference implementation for the `NXP OM13098 development board <https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/general-purpose-mcus/lpc54000-cortex-m4-/lpcxpresso54628-development-board:OM13098>`_ with SystemView support is available `here <https://developer.microej.com/packages/referenceimplementations/U3OER/2.0.1/OM13098-U3OER-fullPackaging-eval-2.0.1.zip>`_. Please contact MicroEJ Support for more information about how to integrate this Platform module.

Reviewer: `events sequence and duration`

Logging library (ej.library.eclasspath.logging)
-----------------------------------------------

> Library ``ej.library.eclasspath.logging`` is based over the ``java.util.logging`` library and follows the same principles:

Reviewer: Tell that this solution is entirely based on Strings (unlike the other solutions).

-  There is one instance of ``LogManager`` by application that manages the hierarchy of loggers.
-  Find or create ``Logger`` objects using the method ``Logger.getLogger(String ID)``. If a logger has already been created with the same name, this logger is returned, otherwise a new logger is created. 

> -  Each ``Logger`` created with this method is registered in the ``LoggerManager`` and can be retrieved using its String ``ID``.

Reviewer: `LogManager` instead of `LoggerManager`

-  You can associate a minimum level to this ``Logger`` so that only messages that have at least this level are logged. The standard levels are listed in the class ``java.util.logging.Level``.
-  The ``Logger`` API provides multiple methods for logging:

>     -  ``log(...)`` methods, that send a ``LogRecord`` with the level argument to the registered ``Handler`` instances.

Reviewer: - remove `,`
- `with a level` or `with a specified level`

>     -  ``log(...)`` methods, that send a ``LogRecord`` with the level argument to the registered ``Handler`` instances.

Reviewer: - remove `,`
- `with a level` or `with a specified level`

Let's see how to use it on our short snippet:

#. Add the following dependency your ``module.ivy``: ``<dependency org="ej.library.eclasspath" name="logging" rev="x.y.z"/>``

#. Call the logging API to log some info text:

   .. code-block:: java
     
      public static void switchState(ApplicationState newState) {
         oldState = currentState;
         currentState = newState;

         Logger logger = Logger.getLogger(Main.class.getName());
         logger.log(Level.INFO, "The application state has changed from " + oldState.toString() + " to "
               + currentState.toString() + ".");
      }

This will produce the following output: 

   .. code-block:: java
      
      main INFO: The application state has changed from UNINSTALLED to INSTALLED.

Message library (ej.library.runtime.message)
--------------------------------------------

Library ``ej.library.runtime.message`` was designed to enable logging while minimizing RAM/ROM footprint and CPU usage. For doing so, it favors the use of integer over strings.

Principles:

- The ``MessageLogger`` type allows for logging messages solely based on integers that identify the message content.

> - Log a message by using methods ``MessageLogger.log(...)``, specifying the log level, the message category and message integer identifer.

Reviewer: `identifier` (missing `i`)

  Use optional arguments to add any useful information to your log such as a throwable or contextual data.
- Log levels are very similar to those of the Logging library. You will find the full level definition in type ``ej.util.message.Level``.
- Combined with the category, the integer ID will allow the user to find the corresponding error/warning/info description.
- Loggers rely on the ``MessageBuilder`` type for message creation. 
  The messages constructed by the ``BasicMessageBuilder`` follow this pattern: `[category]:[LEVEL]=[id]`. The builder will append specified ``Object`` arguments (if any) separated by spaces, then the full stack trace of a throwable (if any).

> - As the ID of the message is an integer, making the output not very human-readable, it is wise to maintain a documentation that describes all message IDs.

Reviewer: `all IDs.`

Usage example:

#. To use this library, add this dependency line in the project module.ivy:

>    ``<dependency org="ej.library.runtime" name="message" rev="x.y.z"/>``

Reviewer: I think that the version used for the tutorial can be put here. It is tedious to have to search for the right version.

Reviewer: (The same goes for the 3 solutions.)

#. Call the message API to log some info:
   
   .. code-block:: java 

      public static void switchState(ApplicationState newState) {
         oldState = currentState;
         currentState = newState;

>          BasicMessageLogger.INSTANCE.log(Level.INFO, "Application", 2, oldState, currentState);

Reviewer: Use constants for the category and the ID.

             :align: center

Follow same principle as before:

      .. code-block:: java 

         public static void switchState(ApplicationState newState) {
            if(Constants.getBoolean(Tracer.TRACE_ENABLED_CONSTANT_PROPERTY)) {
               tracer.recordEvent(0);
            }

            oldState = currentState;
            currentState = newState;

            if(Constants.getBoolean(Tracer.TRACE_ENABLED_CONSTANT_PROPERTY)) {
               tracer.recordEventEnd(0);
            }
         }

      

Use ProGuard
~~~~~~~~~~~~

`ProGuard <https://www.guardsquare.com/en/products/proguard>`_ is a command-line tool that shrinks, optimizes and obfuscates Java code.

It will optimize bytecode as well as detect and remove unused instructions. Therefore it can be used to remove log messages in a production binary.
   
`MicroEJ Github <https://github.com/MicroEJ/>`_ provides a dedicated How-To showing how to `get started with ProGuard <https://github.com/MicroEJ/How-To/tree/1.8.3/Proguard-Get-Started>`_ and remove elements of code from the Logging library (ej.library.eclasspath.logging).

Congratulations!

At this point of the tutorial:

> * You can add logging to your MicroEJ applications while meeting the constraints of embedded applications

Reviewer: Add `.` at the end.

Intended Audience
-----------------

This document's intended audience is engineers who are in charge of adding logs and traces to MicroEJ applications and libraries.

It should also be of interest to all developers looking for best practices when coming to log messages while keeping a low ROM footprint and good overall performances.

Introduction
------------

One straightforward way to add traces in an application is to use the Java basic print methods: ``System.out.println(...)``. 
However, this is not desirable when writing production-quality code, where one typically wants to adjust the log level and provide useful well-formed data and metadata.

Having logs and traces in an application, though, means more code is embedded and executed.
This can impact the overall performances of the application by using more CPU usage at runtime. It also increases the ROM footprint of the application.

Prerequisites
-------------

*  `MicroEJ SDK <https://developer.microej.com/get-started/>`_ ``4.1.5`` or higher.
*  Java Development Kit (JDK) ``1.8.x``.

Overview
--------

In the MicroEJ environment, there are three ways for logging in an application: 
   

> - Trace API
> - Logging library
> - Messages library

Reviewer 2: ```suggestion
- Trace API: a real-time event recording library designed for performance and interaction analysis.
- Logging library: a complete and highly configurable standard logging library.
- Messages library: a lightweight and simple logging library. 
```

This tutorial refactors the following code snippet to illustrate how to log traces with three different libraries:

.. code-block:: java

      public class Main {

         enum ApplicationState {
            INSTALLED, STARTED, STOPPED, UNINSTALLED
         }

         private static ApplicationState currentState;
         private static ApplicationState previousState;

         public static void main(String[] args) {
            currentState = ApplicationState.UNINSTALLED;
            switchState(ApplicationState.INSTALLED);
         }

         public static void switchState(ApplicationState newState) {
            previousState = currentState;
            currentState = newState;
         }
      }

The last section describes how to reduce the footprint when the logs are disabled.

> Trace API (ej.api.trace)

Reviewer 3: By default the traces are disabled. It is already explained how to enable traces below but it could be mentioned here.

Reviewer 4: pulled up the option activation into the trace section

Add logging to MicroEJ applications
===================================

> This document explains how to add logging and tracing to MicroEJ applications and libraries.
> 
> This tutorial presents several solutions. The aim is to help developers to report precise execution context for further debugging and monitoring.

Reviewer 2: ```suggestion
This document explains how to add logging and tracing to MicroEJ applications and libraries with three different solutions. The aim is to help developers to report precise execution context for further debugging and monitoring.
```
?

--------------------------------------------

The library ``ej.library.runtime.message`` was designed to enable logging while minimizing RAM/ROM footprint and CPU usage. For doing so, it favors the use of integer over strings.

Principles:

- The ``MessageLogger`` type allows for logging messages solely based on integers that identify the message content.
- Log a message by using methods ``MessageLogger.log(...)``, specifying the log level, the message category and message integer identifier.
  Use optional arguments to add any useful information to the log such as a throwable or contextual data.
- Log levels are very similar to those of the Logging library. The standard levels are listed in the class ``ej.util.message.Level``.
- Combined with the category, the integer ID allows the user to find the corresponding error/warning/info description.
- Loggers rely on the ``MessageBuilder`` type for message creation. 
  The messages constructed by the ``BasicMessageBuilder`` follow this pattern: `[category]:[LEVEL]=[id]`. The builder appends the specified ``Object`` arguments (if any) separated by spaces, then the full stack trace of a throwable (if any).

> - As the ID of the message is an integer, making the output not very human-readable, it is wise to maintain a documentation that describes all IDs.

Reviewer 2: ```suggestion
- As the ID of the message is an integer, making the output not very human-readable, it is wise to maintain documentation that describes all IDs.
```

Usage example:

#. To use this library, add this dependency line in the ``module.ivy``: ``<dependency org="ej.library.runtime" name="message" rev="2.1.0"/>``

#. Call the message API to log some info:
   
   .. code-block:: java 

      private static final String LOG_CATEGORY = "Application";

      private static final int LOG_ID = 2;

      public static void switchState(ApplicationState newState) {
         previousState = currentState;
         currentState = newState;

         BasicMessageLogger.INSTANCE.log(Level.INFO, LOG_CATEGORY, LOG_ID, previousState, currentState);
      }     

This produces the following output:

   .. code-block:: java
      
      Application:I=2 UNINSTALLED INSTALLED

Remove traces for the production binary
---------------------------------------

There are multiple options for removing all logs and traces when building the production binary.

Wrap logging statements with a check against a static variable
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 

> A boolean constant declared in a ``if`` statement can be used to fully remove portions of code: when this boolean constant is evaluated as ``false``, the wrapped code becomes unreachable and, thus, is not be embedded.

Reviewer 2: ```suggestion
A boolean constant declared in an ``if`` statement can be used to fully remove portions of code: when this boolean constant is evaluated as ``false``, the wrapped code becomes unreachable and, thus, is not embedded.
```

.. note::
    More information about the usage of constants and ``if`` code removal can be found in the :ref:`Classpath <if_constant_removal>` section of the :ref:`Application Developer Guide <application-developer-guide>`.

> #. Let's consider a constant ``com.mycompany.logging`` that we declared as ``false`` in a resource file named ``example.constants.list``.

Reviewer 2: ```suggestion
#. Let's consider a constant ``com.mycompany.logging`` declared as ``false`` in a resource file named ``example.constants.list``.
```

    .. image:: images/tuto_microej_trace_constant.png
        :align: center

#. Add an ``if`` code removal statement to remove a logging, as follows:
   
   .. code-block:: java 

      private static final String LOG_PROPERTY = "com.mycompany.logging";

      public static void switchState(ApplicationState newState) {
         previousState = currentState;
         currentState = newState;

         if (Constants.getBoolean(LOG_PROPERTY)) {
            Logger logger = Logger.getLogger(Main.class.getName());
            logger.log(Level.INFO, "The application state has changed from " + previousState.toString() + " to "
               + currentState.toString() + ".");
         }
      }

> When using the Trace API (ej.api.trace), you can evaluate the value of constant ``Tracer.TRACE_ENABLED_CONSTANT_PROPERTY`` that represents property ``core.trace.enabled``.

Reviewer 2: ```suggestion
When using the Trace API (``ej.api.trace``), you can evaluate the value of the static field ``Tracer.TRACE_ENABLED_CONSTANT_PROPERTY`` that represents the constant ``core.trace.enabled``.
```

> The value of this property can be modified by going to :guilabel:`Launch` > :guilabel:`Launch configurations` then in the tab :guilabel:`Configuration` > :guilabel:`Runtime`, check/uncheck the option :guilabel:`Enable execution traces` to respectively set the value to ``true``/``false``.

Reviewer 2: ```suggestion
The value of this constant can be modified by going to :guilabel:`Launch` > :guilabel:`Launch configurations` then in the tab :guilabel:`Configuration` > :guilabel:`Runtime`, check/uncheck the option :guilabel:`Enable execution traces` to respectively set the value to ``true``/``false``.
```

         .. image:: images/tuto_microej_trace_property.png
             :align: center

> Follow same principle as before:

Reviewer 2: ```suggestion
Follow the same principle as before:
```

      .. code-block:: java 

         private static final int EVENT_ID = 0;

         public static void switchState(ApplicationState newState) {
            if (Constants.getBoolean(Tracer.TRACE_ENABLED_CONSTANT_PROPERTY)) {
               tracer.recordEvent(EVENT_ID);
            }

            previousState = currentState;
            currentState = newState;

            if (Constants.getBoolean(Tracer.TRACE_ENABLED_CONSTANT_PROPERTY)) {
               tracer.recordEventEnd(EVENT_ID);
            }
         }

      

Use ProGuard
~~~~~~~~~~~~

> `ProGuard <https://www.guardsquare.com/en/products/proguard>`_ is a command-line tool that shrinks, optimizes and obfuscates Java code.

Reviewer 2: ```suggestion
`ProGuard <https://www.guardsquare.com/en/products/proguard>`_ is a command-line tool that shrinks, optimizes, and obfuscates Java code.
```

> It optimizes bytecode as well as detect and remove unused instructions. Therefore it can be used to remove log messages in a production binary.

Reviewer 2: ```suggestion
It optimizes bytecode as well as it detects and removes unused instructions. Therefore it can be used to remove log messages in a production binary.
```

   
`MicroEJ Github <https://github.com/MicroEJ/>`_ provides a dedicated How-To showing how to `get started with ProGuard <https://github.com/MicroEJ/How-To/tree/1.8.3/Proguard-Get-Started>`_ and remove elements of code from the Logging library (ej.library.eclasspath.logging).

Congratulations!

> At this point of the tutorial:

Reviewer 2: ```suggestion
At this point in the tutorial:
```
